lib/posix/pthread_mutex.c

```diff
@@ -90,60 +91,62 @@ struct posix_mutex *to_posix_mutex(pthread_mutex_t *mu)
 	/* Initialize the posix_mutex */
 	m = &posix_mutex_pool[bit];
 
-	m->owner = NULL;
-	m->lock_count = 0U;
-
-	z_waitq_init(&m->wait_q);
+	err = k_mutex_init(m);
+	__ASSERT_NO_MSG(err == 0);
 
 	return m;
 }
 
 static int acquire_mutex(pthread_mutex_t *mu, k_timeout_t timeout)
 {
-	int rc = 0;
+	int type;
+	size_t bit;
+	int ret = 0;
+	struct k_mutex *m;
 	k_spinlock_key_t key;
-	struct posix_mutex *m;
-
-	key = k_spin_lock(&z_pthread_spinlock);
 
 	m = to_posix_mutex(mu);
 	if (m == NULL) {
-		k_spin_unlock(&z_pthread_spinlock, key);
 		return EINVAL;
 	}
 
-	if (m->lock_count == 0U && m->owner == NULL) {
-		m->lock_count++;
-		m->owner = k_current_get();
-
-		k_spin_unlock(&z_pthread_spinlock, key);
-		return 0;
-	} else if (m->owner == k_current_get()) {
-		if (m->type == PTHREAD_MUTEX_RECURSIVE &&
-		    m->lock_count < MUTEX_MAX_REC_LOCK) {
-			m->lock_count++;
-			rc = 0;
-		} else if (m->type == PTHREAD_MUTEX_ERRORCHECK) {
-			rc = EDEADLK;
-		} else {
-			rc = EINVAL;
+	bit = posix_mutex_to_offset(m);
+	type = posix_mutex_type[bit];
+
+	key = k_spin_lock(&pthread_mutex_spinlock);
+	if (m->owner == k_current_get()) {
+		switch (type) {
+		case PTHREAD_MUTEX_NORMAL:
+			if (K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
+				ret = EBUSY;
+				break;
+			}
+			/* On most POSIX systems, this usually results in an infinite loop */
+			do {
+				k_yield();
+			} while (true);
```

Comment:
Hmm, as per previous explanation I understand that we expect current posix thread to hang in this case.

Do we expect that it will also break other threads (as we hold `pthread_mutex_spinlock` spinlock and never unlock it)?

Reply:
Probably we should unlock spinlock and call k_sleep(K_FOREVER) here.

Reply:
Sure